### PR TITLE
Improve TestPipeline AppName detection

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

After encountering the TestPipeline portion of the stack, try to find
a method annotated with the org.junit.Test annotation. If such a method
is found, return it as the StackTraceElement of the calling class.

This improves TestPipeline name detection for RunnableOnService tests to
avoid duplication

### CODE REVIEW

**Comment by tgroh** [comment]
R: @lukecwik

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipeline.java [hunk 1]
```diff
@@ -183,12 +184,27 @@ private static String getAppName() {
...
       if (!TestPipeline.class.getName().equals(next.getClassName())) {
-        return Optional.of(next);
+        if (candidate == null) {
+          candidate = next;
+        }
+        try {
+          Class<?> nextClass = Class.forName(next.getClassName());
+          for (Method method : nextClass.getMethods()) {
+            if (method.getName().equals(next.getMethodName())) {
+              if (method.getAnnotation(org.junit.Test.class) != null) {
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipeline.java]: https://docs.oracle.com/javase/7/docs/api/java/lang/reflect/AccessibleObject.html#isAnnotationPresent(java.lang.Class)

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipeline.java [hunk 2]
```diff
@@ -183,12 +184,27 @@ private static String getAppName() {
       }
     }
     // Then find the first instance after that is not the TestPipeline
+    StackTraceElement candidate = null;
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipeline.java]: Optional<StackTraceElement>

candidate -> firstClassBeforeTestPipeline

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/testing/TestPipelineTest.java [hunk 3]
```diff
@@ -91,4 +91,18 @@ public void testCreationOfPipelineOptionsFromReallyVerboselyNamedTestCase() thro
   public void testToString() {
     assertEquals("TestPipeline#TestPipelineTest-testToString", TestPipeline.create().toString());
   }
+
+  @Test
+  public void testToStringNestedMethod() {
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/testing/TestPipelineTest.java]: Add a test for the nested class case:
TestClass.myTest
TestClass.MyTestClassHelper.runPipeline
TestPipeline.run

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipeline.java (hunk 1 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipeline.java]: Done

File: sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipeline.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/main/java/com/google/cloud/dataflow/sdk/testing/TestPipeline.java]: Done

File: sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/testing/TestPipelineTest.java (hunk 3 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/com/google/cloud/dataflow/sdk/testing/TestPipelineTest.java]: Done

**Comment by lukecwik** [comment]
LGTM

**Pull request closed without merging**